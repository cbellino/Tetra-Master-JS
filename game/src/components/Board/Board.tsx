import * as React from "react";

import { compose, defaultProps } from "recompose";
import { Grid, Vector2 } from "../../models";
import { EnhancedBoardCell } from "./Board.Cell";
import "./Board.css";

type EnhancedBoardProps = {
  grid: Grid;
  onCellClick?: (position: Vector2) => void;
};
type BoardProps = {
  grid: Grid;
  onCellClick: (position: Vector2) => void;
};

const Board: React.SFC<BoardProps> = ({ grid, onCellClick }) => (
  <div className="board">
    <div className="grid">
      {grid.map((row, x) =>
        row.map((cell, y) => (
          <EnhancedBoardCell
            key={`${x}-${y}`}
            cell={cell}
            position={new Vector2(x, y)}
            onClick={onCellClick}
          />
        )),
      )}
    </div>
  </div>
);

const withDefaultProps = defaultProps({
  onCellClick: () => {},
});

const enhance = compose<BoardProps, EnhancedBoardProps>(withDefaultProps);

export const EnhancedBoard = enhance(Board);
