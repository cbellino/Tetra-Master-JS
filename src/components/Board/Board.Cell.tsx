import * as React from "react";
import { defaultProps } from "recompose";

import { Cell, Vector2 } from "../../models";
import TileCard from "../TileCard";

type CellProps = {
  position: Vector2;
  cell: Cell;
  onClick: (position: Vector2) => void;
};

const onClickWithPosition = (onClick, position) => clickEvent => {
  onClick(position);
};

const BoardCell: React.SFC<CellProps> = ({ position, cell, onClick }) => (
  <div
    className={`board-cell position-${position.x}-${position.y}`}
    role="button"
    onClick={onClickWithPosition(onClick, position)}
  >
    {cell.tileId && <TileCard tileId={cell.tileId} playerId={cell.playerId} />}
  </div>
);

export const EnhancedBoardCell = defaultProps({
  onClick: _ => {},
})(BoardCell);
