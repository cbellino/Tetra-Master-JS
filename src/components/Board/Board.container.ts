import { connect } from "react-redux";

import { getBoardGrid, placeCurrentPlayerTile } from "../../store";
import { finishTurn } from "../../store/turn/turn.actions";
import { Board } from "./Board";

const mapStateToProps = state => ({
  grid: getBoardGrid(state),
});

const handleCellClick = dispatch => position => {
  dispatch(placeCurrentPlayerTile(position));
  dispatch(finishTurn());
};

const mapDispatchToProps = dispatch => ({
  onCellClick: handleCellClick(dispatch),
});

const enhance = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default enhance(Board);
