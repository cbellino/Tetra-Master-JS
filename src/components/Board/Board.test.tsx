import { shallow } from "enzyme";
import * as React from "react";

import { EnhancedBoard as Board } from "./Board";
import { EnhancedBoardCell } from "./Board.Cell";

const mockGrid = [];
const mockGridWithCells = [
  [{ tileId: "1", playerId: "1" }, { tileId: "1", playerId: "1" }],
  [{ tileId: "1", playerId: "1" }, { tileId: "1", playerId: "1" }],
];

describe("Board", () => {
  describe("when the grid is empty", () => {
    it("should not display any cell", () => {
      const wrapper = shallow(<Board grid={mockGrid} />).dive();

      expect(wrapper.find(EnhancedBoardCell)).toHaveLength(0);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("should have a className containing the player id", () => {
    const wrapper = shallow(<Board grid={mockGridWithCells} />).dive();

    expect(wrapper.find(EnhancedBoardCell)).toHaveLength(4);
    expect(wrapper).toMatchSnapshot();
  });
});
