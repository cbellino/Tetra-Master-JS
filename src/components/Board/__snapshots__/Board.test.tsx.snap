// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should have a className containing the player id 1`] = `
<div
  className="board"
>
  <div
    className="grid"
  >
    <defaultProps(BoardCell)
      cell={
        Object {
          "playerId": "1",
          "tileId": "1",
        }
      }
      key="0-0"
      onClick={[Function]}
      position={
        Vector2 {
          "x": 0,
          "y": 0,
        }
      }
    />
    <defaultProps(BoardCell)
      cell={
        Object {
          "playerId": "1",
          "tileId": "1",
        }
      }
      key="0-1"
      onClick={[Function]}
      position={
        Vector2 {
          "x": 0,
          "y": 1,
        }
      }
    />
    <defaultProps(BoardCell)
      cell={
        Object {
          "playerId": "1",
          "tileId": "1",
        }
      }
      key="1-0"
      onClick={[Function]}
      position={
        Vector2 {
          "x": 1,
          "y": 0,
        }
      }
    />
    <defaultProps(BoardCell)
      cell={
        Object {
          "playerId": "1",
          "tileId": "1",
        }
      }
      key="1-1"
      onClick={[Function]}
      position={
        Vector2 {
          "x": 1,
          "y": 1,
        }
      }
    />
  </div>
</div>
`;

exports[`Board when the grid is empty should not display any cell 1`] = `
<div
  className="board"
>
  <div
    className="grid"
  />
</div>
`;
